services/graph/pkg/service/v0/api_driveitem_permissions_links.go

```diff
@@ -0,0 +1,425 @@
+			return libregraph.Permission{}, errorcode.New(errorcode.InvalidRequest, "invalid expiration date")
+		}
+		req.GetGrant().Expiration = expireTime
+	}
+
+	// set displayname and password protected as arbitrary metadata
+	req.ResourceInfo.ArbitraryMetadata = &providerv1beta1.ArbitraryMetadata{
+		Metadata: map[string]string{
+			"name":      createLink.GetDisplayName(),
+			"quicklink": strconv.FormatBool(createLink.GetLibreGraphQuickLink()),
+		},
+	}
+	createResp, err := gatewayClient.CreatePublicShare(ctx, &req)
+	if err != nil {
+		s.logger.Error().Err(err).Msg("transport error, could not create link")
+		return libregraph.Permission{}, errorcode.New(errorcode.GeneralException, err.Error())
+	}
+	if statusCode := createResp.GetStatus().GetCode(); statusCode != rpc.Code_CODE_OK {
+		return libregraph.Permission{}, errorcode.New(cs3StatusToErrCode(statusCode), createResp.Status.Message)
+	}
+	link := createResp.GetShare()
+	perm, err := s.libreGraphPermissionFromCS3PublicShare(link)
+	if err != nil {
+		return libregraph.Permission{}, errorcode.New(errorcode.GeneralException, err.Error())
+	}
+	return *perm, nil
+}
+
+func (s DriveItemPermissionsService) CreateSpaceRootLink(ctx context.Context, driveID storageprovider.ResourceId, createLink libregraph.DriveItemCreateLink) (libregraph.Permission, error) {
+	gatewayClient, err := s.gatewaySelector.Next()
+	if err != nil {
+		return libregraph.Permission{}, err
+	}
+	space, err := utils.GetSpace(ctx, storagespace.FormatResourceID(driveID), gatewayClient)
+	if err != nil {
+		return libregraph.Permission{}, err
+	}
+
+	if space.SpaceType != _spaceTypeProject {
+		return libregraph.Permission{}, errorcode.New(errorcode.InvalidRequest, "unsupported space type")
+	}
+
+	rootResourceID := space.GetRoot()
+	return s.CreateLink(ctx, *rootResourceID, createLink)
+}
+
+func (s DriveItemPermissionsService) SetPublicLinkPassword(ctx context.Context, driveItemId storageprovider.ResourceId, permissionID string, password string) (libregraph.Permission, error) {
+	publicShare, err := s.getCS3PublicShareByID(ctx, permissionID)
+	if err != nil {
+		return libregraph.Permission{}, err
+	}
+
+	// The resourceID of the shared resource need to match the item ID from the Request Path
+	// otherwise this is an invalid Request.
+	if !utils.ResourceIDEqual(publicShare.GetResourceId(), &driveItemId) {
+		s.logger.Debug().Msg("resourceID of shared does not match itemID")
+		return libregraph.Permission{}, errorcode.New(errorcode.InvalidRequest, "permissionID and itemID do not match")
+	}
+
+	permission, err := s.updatePublicLinkPassword(ctx, permissionID, password)
+	if err != nil {
+		return libregraph.Permission{}, err
+	}
+	return *permission, nil
+}
+
+func (s DriveItemPermissionsService) SetPublicLinkPasswordOnSpaceRoot(ctx context.Context, driveID storageprovider.ResourceId, permissionID string, password string) (libregraph.Permission, error) {
```

Comment:
exported func `SetPublicLinkPasswordOnSpaceRoot` should have a comment or be unexported